File: openai-tool-server/test_server.py
"""
Test script for the Zest OpenAI Tool Server
"""

import httpx
import json
import time
from typing import Dict, Any


def test_endpoint(client: httpx.Client, method: str, endpoint: str, 
                 json_data: Dict[str, Any] = None, expected_status: int = 200):
    """Test a single endpoint"""
    print(f"\n🧪 Testing {method} {endpoint}")
    
    try:
        if method == "GET":
            response = client.get(endpoint)
        elif method == "POST":
            response = client.post(endpoint, json=json_data)
        else:
            raise ValueError(f"Unsupported method: {method}")
        
        print(f"   Status: {response.status_code} {'✅' if response.status_code == expected_status else '❌'}")
        
        if response.status_code == 200:
            data = response.json()
            print(f"   Response: {json.dumps(data, indent=2)[:200]}...")
            return True, data
        else:
            print(f"   Error: {response.text}")
            return False, None
            
    except Exception as e:
        print(f"   ❌ Exception: {str(e)}")
        return False, None


def run_tests(base_url: str = "http://localhost:8000", api_key: str = ""):
    """Run all tests"""
    
    print("🚀 Starting Zest OpenAI Tool Server Tests")
    print(f"   Server: {base_url}")
    print(f"   API Key: {'Set' if api_key else 'Not set'}")
    
    # Create client
    headers = {}
    if api_key:
        headers["Authorization"] = f"Bearer {api_key}"
    
    client = httpx.Client(base_url=base_url, headers=headers, timeout=30.0)
    
    try:
        # Test health endpoint
        success, health = test_endpoint(client, "GET", "/health")
        if not success:
            print("\n❌ Server not responding. Is it running?")
            return
        
        proxy_connected = health.get("proxy_connected", False)
        if not proxy_connected:
            print("\n⚠️  Warning: Proxy not connected. Some tests may fail.")
        
        # Test OpenAPI schema
        test_endpoint(client, "GET", "/openapi.json")
        
        # Test tools listing
        test_endpoint(client, "GET", "/tools")
        
        # Test status
        test_endpoint(client, "GET", "/status")
        
        # Test config
        test_endpoint(client, "GET", "/config")
        
        if proxy_connected:
            # Test explore
            test_endpoint(client, "POST", "/explore", {
                "query": "What programming languages are used in this project?",
                "generate_report": False
            })
            
            # Test search
            test_endpoint(client, "POST", "/search", {
                "query": "main method",
                "max_results": 3
            })
            
            # Test find by name
            test_endpoint(client, "POST", "/find-by-name", {
                "name": "Main",
                "type": "class"
            })
            
            # Test augment
            test_endpoint(client, "POST", "/augment", {
                "query": "How does authentication work?"
            })
            
            # Test execute tool
            test_endpoint(client, "POST", "/execute-tool", {
                "tool": "list_files",
                "parameters": {"directory": "/"}
            })
        
        print("\n✅ All tests completed!")
        
    finally:
        client.close()


def performance_test(base_url: str = "http://localhost:8000", api_key: str = ""):
    """Run performance tests"""
    
    print("\n⚡ Running Performance Tests")
    
    headers = {}
    if api_key:
        headers["Authorization"] = f"Bearer {api_key}"
    
    with httpx.Client(base_url=base_url, headers=headers, timeout=300.0) as client:
        # Test concurrent requests
        print("\n📊 Testing concurrent requests...")
        
        queries = [
            "What is the main purpose of this application?",
            "Find all database models",
            "List all API endpoints",
            "What design patterns are used?",
            "Find security-related code"
        ]
        
        start_time = time.time()
        
        for i, query in enumerate(queries):
            print(f"   Query {i+1}: {query[:50]}...")
            response = client.post("/search", json={"query": query, "max_results": 5})
            print(f"   Status: {response.status_code}")
        
        elapsed = time.time() - start_time
        print(f"\n   Total time: {elapsed:.2f}s")
        print(f"   Average per query: {elapsed/len(queries):.2f}s")


if __name__ == "__main__":
    import argparse
    
    parser = argparse.ArgumentParser(description="Test Zest OpenAI Tool Server")
    parser.add_argument("--url", default="http://localhost:8000", help="Server URL")
    parser.add_argument("--api-key", default="", help="API key for authentication")
    parser.add_argument("--performance", action="store_true", help="Run performance tests")
    
    args = parser.parse_args()
    
    # Run basic tests
    run_tests(args.url, args.api_key)
    
    # Run performance tests if requested
    if args.performance:
        performance_test(args.url, args.api_key)

File: openai-tool-server-js/test-server.js
import fetch from 'node-fetch';
import { performance } from 'perf_hooks';

/**
 * Test suite for Zest OpenAI Tool Server
 */
class ServerTester {
  constructor(baseUrl = 'http://localhost:8000', apiKey = '') {
    this.baseUrl = baseUrl;
    this.headers = {
      'Content-Type': 'application/json'
    };
    
    if (apiKey) {
      this.headers['Authorization'] = `Bearer ${apiKey}`;
    }
    
    this.testResults = [];
  }

  async testEndpoint(method, endpoint, data = null, expectedStatus = 200) {
    const testName = `${method} ${endpoint}`;
    console.log(`\n🧪 Testing ${testName}`);
    
    const startTime = performance.now();
    
    try {
      const options = {
        method,
        headers: this.headers
      };
      
      if (data && method !== 'GET') {
        options.body = JSON.stringify(data);
      }
      
      const response = await fetch(`${this.baseUrl}${endpoint}`, options);
      const responseTime = performance.now() - startTime;
      
      const success = response.status === expectedStatus;
      console.log(`   Status: ${response.status} ${success ? '✅' : '❌'} (${responseTime.toFixed(2)}ms)`);
      
      let responseData = null;
      if (response.ok) {
        responseData = await response.json();
        console.log(`   Response: ${JSON.stringify(responseData).substring(0, 100)}...`);
      } else {
        const errorText = await response.text();
        console.log(`   Error: ${errorText}`);
      }
      
      this.testResults.push({
        name: testName,
        success,
        status: response.status,
        responseTime,
        data: responseData
      });
      
      return { success, data: responseData };
      
    } catch (error) {
      console.log(`   ❌ Exception: ${error.message}`);
      this.testResults.push({
        name: testName,
        success: false,
        error: error.message,
        responseTime: performance.now() - startTime
      });
      return { success: false, error: error.message };
    }
  }

  async runBasicTests() {
    console.log('🏃 Running Basic Tests\n');
    
    // Health check
    const health = await this.testEndpoint('GET', '/health');
    if (!health.success) {
      console.log('\n❌ Server not responding. Is it running?');
      return false;
    }
    
    const proxyConnected = health.data?.proxy_connected || false;
    if (!proxyConnected) {
      console.log('\n⚠️  Warning: Proxy not connected. Some tests may fail.');
    }
    
    // Test OpenAPI schema
    await this.testEndpoint('GET', '/openapi.json');
    
    // Test Swagger docs
    await this.testEndpoint('GET', '/docs', null, 200);
    
    // Test tools listing
    await this.testEndpoint('GET', '/tools');
    
    // Test status
    await this.testEndpoint('GET', '/status');
    
    // Test config
    await this.testEndpoint('GET', '/config');
    
    return proxyConnected;
  }

  async runProxyTests() {
    console.log('\n🏃 Running Proxy-Dependent Tests\n');
    
    // Test explore
    await this.testEndpoint('POST', '/explore', {
      query: 'What programming languages are used in this project?',
      generate_report: false
    });
    
    // Test search
    await this.testEndpoint('POST', '/search', {
      query: 'main method',
      max_results: 3
    });
    
    // Test find by name
    await this.testEndpoint('POST', '/find-by-name', {
      name: 'Main',
      type: 'class'
    });
    
    // Test augment
    await this.testEndpoint('POST', '/augment', {
      query: 'How does authentication work?'
    });
    
    // Test execute tool
    await this.testEndpoint('POST', '/execute-tool', {
      tool: 'list_files',
      parameters: { directory: '/' }
    });
    
    // Test read file (may fail if file doesn't exist)
    await this.testEndpoint('POST', '/read-file', {
      file_path: 'README.md'
    });
    
    // Test find relationships
    await this.testEndpoint('POST', '/find-relationships', {
      element_id: 'com.example.Main'
    });
    
    // Test find usages
    await this.testEndpoint('POST', '/find-usages', {
      element_id: 'com.example.Main'
    });
    
    // Test class info
    await this.testEndpoint('POST', '/class-info', {
      class_name: 'com.example.Main'
    });
  }

  async runValidationTests() {
    console.log('\n🏃 Running Validation Tests\n');
    
    // Test missing required fields
    await this.testEndpoint('POST', '/explore', {}, 400);
    
    // Test invalid field types
    await this.testEndpoint('POST', '/search', {
      query: 'test',
      max_results: 'invalid'
    }, 400);
    
    // Test invalid enum values
    await this.testEndpoint('POST', '/find-by-name', {
      name: 'Test',
      type: 'invalid_type'
    }, 400);
  }

  async runAuthTests(wrongApiKey = 'wrong-key') {
    console.log('\n🏃 Running Authentication Tests\n');
    
    // Save original headers
    const originalHeaders = { ...this.headers };
    
    // Test with wrong API key
    this.headers['Authorization'] = `Bearer ${wrongApiKey}`;
    await this.testEndpoint('GET', '/tools', null, 401);
    
    // Test without API key
    delete this.headers['Authorization'];
    await this.testEndpoint('GET', '/tools', null, 401);
    
    // Test with malformed auth header
    this.headers['Authorization'] = 'InvalidFormat';
    await this.testEndpoint('GET', '/tools', null, 401);
    
    // Restore original headers
    this.headers = originalHeaders;
  }

  async runPerformanceTests() {
    console.log('\n⚡ Running Performance Tests\n');
    
    const queries = [
      'What is the main purpose of this application?',
      'Find all database models',
      'List all API endpoints',
      'What design patterns are used?',
      'Find security-related code'
    ];
    
    console.log('📊 Testing concurrent requests...');
    const startTime = performance.now();
    
    const promises = queries.map((query, i) => {
      console.log(`   Query ${i + 1}: ${query.substring(0, 50)}...`);
      return this.testEndpoint('POST', '/search', {
        query,
        max_results: 5
      });
    });
    
    await Promise.all(promises);
    
    const totalTime = performance.now() - startTime;
    console.log(`\n   Total time: ${(totalTime / 1000).toFixed(2)}s`);
    console.log(`   Average per query: ${(totalTime / queries.length / 1000).toFixed(2)}s`);
  }

  generateReport() {
    console.log('\n📊 Test Report\n');
    console.log('='.repeat(60));
    
    const totalTests = this.testResults.length;
    const passedTests = this.testResults.filter(t => t.success).length;
    const failedTests = totalTests - passedTests;
    const avgResponseTime = this.testResults.reduce((sum, t) => sum + t.responseTime, 0) / totalTests;
    
    console.log(`Total Tests: ${totalTests}`);
    console.log(`Passed: ${passedTests} ✅`);
    console.log(`Failed: ${failedTests} ❌`);
    console.log(`Success Rate: ${((passedTests / totalTests) * 100).toFixed(1)}%`);
    console.log(`Average Response Time: ${avgResponseTime.toFixed(2)}ms`);
    
    if (failedTests > 0) {
      console.log('\n❌ Failed Tests:');
      this.testResults
        .filter(t => !t.success)
        .forEach(t => {
          console.log(`  - ${t.name}: ${t.error || `Status ${t.status}`}`);
        });
    }
    
    console.log('\n' + '='.repeat(60));
  }
}

// Main test runner
async function main() {
  console.log('🚀 Zest OpenAI Tool Server Test Suite\n');
  
  // Parse command line arguments
  const args = process.argv.slice(2);
  const url = args.find(arg => arg.startsWith('--url='))?.split('=')[1] || 'http://localhost:8000';
  const apiKey = args.find(arg => arg.startsWith('--api-key='))?.split('=')[1] || '';
  const runPerf = args.includes('--performance');
  const runAuth = args.includes('--auth');
  
  console.log(`Server URL: ${url}`);
  console.log(`API Key: ${apiKey ? 'Set' : 'Not set'}`);
  console.log('');
  
  const tester = new ServerTester(url, apiKey);
  
  try {
    // Run basic tests
    const proxyConnected = await tester.runBasicTests();
    
    // Run validation tests
    await tester.runValidationTests();
    
    // Run auth tests if API key is configured
    if (apiKey && runAuth) {
      await tester.runAuthTests();
    }
    
    // Run proxy-dependent tests if connected
    if (proxyConnected) {
      await tester.runProxyTests();
      
      // Run performance tests if requested
      if (runPerf) {
        await tester.runPerformanceTests();
      }
    }
    
    // Generate report
    tester.generateReport();
    
  } catch (error) {
    console.error('\n❌ Test suite error:', error.message);
  }
}

// Run tests
if (import.meta.url === `file://${process.argv[1]}`) {
  console.log('Usage: node test-server.js [--url=http://localhost:8000] [--api-key=your-key] [--performance] [--auth]\n');
  main().catch(console.error);
}

export { ServerTester };
